rust/lance-table/src/io/commit.rs

```diff
@@ -95,45 +179,67 @@ async fn current_manifest_path(
 
-    let current = manifest_files
-        .objects
-        .into_iter()
-        .filter(|meta| {
-            meta.location.filename().is_some()
-                && meta
-                    .location
-                    .filename()
-                    .unwrap()
-                    .ends_with(MANIFEST_EXTENSION)
-        })
-        .filter_map(|meta| {
-            let version = meta
-                .location
-                .filename()
-                .unwrap()
-                .split_once('.')
-                .and_then(|(version_str, _)| version_str.parse::<u64>().ok())?;
-            Some((version, meta))
-        })
-        .max_by_key(|(version, _)| *version);
-
-    if let Some((version, meta)) = current {
-        Ok(ManifestLocation {
-            version,
-            path: meta.location,
-            size: Some(meta.size as u64),
-        })
-    } else {
-        Err(Error::NotFound {
-            uri: manifest_path(base, 1).to_string(),
+    let mut valid_manifests = manifest_files.try_filter_map(|res| {
+        if let Some(scheme) = ManifestNamingScheme::detect_scheme(res.location.filename().unwrap())
+        {
+            future::ready(Ok(Some((scheme, res))))
+        } else {
+            future::ready(Ok(None))
+        }
+    });
+
+    let first = valid_manifests.next().await.transpose()?;
+    match first {
+        // If the first valid manifest we see is V2, we can assume that we are using
+        // V2 naming scheme for all manifests.
+        Some((scheme @ ManifestNamingScheme::V2, meta)) => {
+            let version = scheme
+                .parse_version(meta.location.filename().unwrap())
+                .unwrap();
+            Ok(ManifestLocation {
+                version,
+                path: meta.location,
+                size: Some(meta.size as u64),
+                naming_scheme: scheme,
+            })
+        }
+        // If the first valid manifest we see if V1, assume for now that we are
+        // using V1 naming scheme for all manifests. Since we are listing the
+        // directory anyways, we will assert there aren't any V2 manifests.
+        Some((scheme @ ManifestNamingScheme::V1, meta)) => {
+            let mut current_version = scheme
+                .parse_version(meta.location.filename().unwrap())
+                .unwrap();
+            let mut current_meta = meta;
+
+            while let Some((scheme, meta)) = valid_manifests.next().await.transpose()? {
+                if matches!(scheme, ManifestNamingScheme::V2) {
+                    return Err(Error::Internal {
+                        message: "Found V2 manifest in a V1 manifest directory".to_string(),
+                        location: location!(),
+                    });
+                }
+                let version = scheme
+                    .parse_version(meta.location.filename().unwrap())
+                    .unwrap();
+                if version > current_version {
```

Comment:
In theory this must be true according to our understanding of object stores yes?

Reply:
`10.manifest` will be before `8.manifest`, so not necessarily. It would be if we had zero-padded the numbers.